### PR TITLE
Early exit for empty operators in pc assembly

### PR DESCRIPTION
**Pull request by jeremylt**

### CODE REVIEW

File: interface/ceed-preconditioning.c [hunk 1]
```diff
@@ -512,7 +512,6 @@ static int CeedSingleOperatorAssemble(CeedOperator op, CeedInt offset, CeedVecto
     return CeedError(ceed, CEED_ERROR_UNSUPPORTED, "Composite operator not supported");
     // LCOV_EXCL_STOP
   }
-  if (op->num_elem == 0) return CEED_ERROR_SUCCESS;
```
Comment (jedbrown) [inline interface/ceed-preconditioning.c:515]: This commit fixes the pbdiagonal case, but regular assembly now fails. I can share the reproducer for this stack or maybe you can see where the fix goes.
```
/home/jed/src/libCEED/interface/ceed-vector.c:401 in CeedVectorGetArray(): CeedVector has no valid data to read, must set data with CeedVectorSetValue or CeedVectorSetArray

Thread 1 "fluids-navierst" received signal SIGABRT, Aborted.
0x00007ffff548f64c in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff548f64c in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff543f958 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff542953d in abort () from /usr/lib/libc.so.6
#3  0x00007ffff58196cf in CeedErrorAbort (ceed=<optimized out>, filename=<optimized out>, line_no=<optimized out>, func=<optimized out>, err_code=<optimized out>, format=0x7ffff598e2b0 "CeedVector has no valid data to read, must set data with CeedVectorSetValue or CeedVectorSetArray", args=0x7fffffff9a00) at /home/jed/src/libCEED/interface/ceed.c:1146
#4  0x00007ffff5819e38 in CeedErrorImpl (ceed=<optimized out>, filename=filename@entry=0x7ffff598e020 "/home/jed/src/libCEED/interface/ceed-vector.c", lineno=lineno@entry=401, func=func@entry=0x7ffff598e9e0 <__func__.12> "CeedVectorGetArray", ecode=ecode@entry=-2, format=format@entry=0x7ffff598e2b0 "CeedVector has no valid data to read, must set data with CeedVectorSetValue or CeedVectorSetArray") at /home/jed/src/libCEED/interface/ceed.c:1085
#5  0x00007ffff582bc17 in CeedVectorGetArray (vec=0x55559b4a5990, mem_type=mem_type@entry=CEED_MEM_DEVICE, array=array@entry=0x7fffffff9bc8) at /home/jed/src/libCEED/interface/ceed-vector.c:401
#6  0x00007ffff58493a2 in CeedOperatorLinearAssembleQFunctionCore_Cuda (op=<optimized out>, build_objects=build_objects@entry=true, assembled=0x7fffffff9d58, rstr=0x7fffffff9d60, request=<optimized out>) at /home/jed/src/libCEED/backends/cuda-ref/ceed-cuda-ref-operator.c:483
#7  0x00007ffff584997d in CeedOperatorLinearAssembleQFunction_Cuda (op=<optimized out>, assembled=<optimized out>, rstr=<optimized out>, request=<optimized out>) at /home/jed/src/libCEED/backends/cuda-ref/ceed-cuda-ref-operator.c:580
#8  0x00007ffff58258c1 in CeedOperatorLinearAssembleQFunctionBuildOrUpdate (op=op@entry=0x55559ad019f0, assembled=assembled@entry=0x7fffffff9e80, rstr=rstr@entry=0x7fffffff9e88, request=0x7ffff5b62928 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1376
#9  0x00007ffff5846681 in CeedSingleOperatorAssemble_Cuda (op=0x55559ad019f0, offset=10342400, values=0x55559af04460) at /home/jed/src/libCEED/backends/cuda-ref/ceed-cuda-ref-operator.c:1087
#10 0x00007ffff5826653 in CeedSingleOperatorAssemble (op=0x55559ad019f0, offset=offset@entry=10342400, values=values@entry=0x55559af04460) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:518
#11 0x00007ffff58266d2 in CeedSingleOperatorAssemble (op=0x55555a4c2c00, offset=offset@entry=10342400, values=values@entry=0x55559af04460) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:526
#12 0x00007ffff582784f in CeedOperatorLinearAssemble (op=<optimized out>, values=values@entry=0x55559af04460) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1792
#13 0x000055555559cedb in FormSetValues (user=user@entry=0x555555c56980, pbdiagonal=<optimized out>, J=J@entry=0x555555c2d570, coo_values=0x55559af04460) at /home/jed/src/libCEED/examples/fluids/src/setupts.c:297
#14 0x000055555559d33b in FormIJacobian_NS (ts=<optimized out>, t=<optimized out>, Q=<optimized out>, Q_dot=<optimized out>, shift=<optimized out>, J=0x55559a0951f0, J_pre=0x555555c2d570, user_data=<optimized out>) at /home/jed/src/libCEED/examples/fluids/src/setupts.c:329
#15 0x00007ffff6e45777 in TSComputeIJacobian (ts=0x55558d570920, t=0.001, U=0x55559a7c2880, Udot=0x55559a7c80d0, shift=1000, A=0x55559a0951f0, B=0x555555c2d570, imex=imex@entry=PETSC_FALSE) at /home/jed/petsc/src/ts/interface/ts.c:893
#16 0x00007ffff6e90430 in SNESTSFormJacobian_Alpha (snes=<optimized out>, X=<optimized out>, J=<optimized out>, P=<optimized out>, ts=<optimized out>) at /home/jed/petsc/src/ts/impls/implicit/alpha/alpha1.c:300
#17 0x00007ffff6e3c95b in SNESTSFormJacobian (snes=<optimized out>, U=<optimized out>, A=<optimized out>, B=<optimized out>, ctx=<optimized out>) at /home/jed/petsc/src/ts/interface/ts.c:4236
#18 0x00007ffff6df8490 in SNESComputeJacobian (snes=snes@entry=0x55558d574060, X=X@entry=0x55559a7ca1b0, A=0x55559a0951f0, B=0x555555c2d570) at /home/jed/petsc/src/snes/interface/snes.c:2805
#19 0x00007ffff6e1cb07 in SNESSolve_NEWTONLS (snes=0x55558d574060) at /home/jed/petsc/src/snes/impls/ls/ls.c:207
#20 0x00007ffff6dfcae5 in SNESSolve (snes=0x55558d574060, b=b@entry=0x0, x=<optimized out>) at /home/jed/petsc/src/snes/interface/snes.c:4693
#21 0x00007ffff6e909c9 in TSAlpha_SNESSolve (ts=0x55558d570920, x=<optimized out>, b=0x0) at /home/jed/petsc/src/ts/impls/implicit/alpha/alpha1.c:82
#22 0x00007ffff6e913e9 in TSAlpha_Restart (initok=0x7fffffffa95c, ts=0x55558d570920) at /home/jed/petsc/src/ts/impls/implicit/alpha/alpha1.c:121
#23 TSStep_Alpha (ts=0x55558d570920) at /home/jed/petsc/src/ts/impls/implicit/alpha/alpha1.c:180
#24 0x00007ffff6e40ce7 in TSStep (ts=0x55558d570920) at /home/jed/petsc/src/ts/interface/ts.c:3425
#25 0x000055555559f778 in TSSolve_NS (dm=<optimized out>, user=<optimized out>, app_ctx=0x555555c3ee60, phys=<optimized out>, Q=Q@entry=0x7fffffffab10, f_time=f_time@entry=0x7fffffffab28, ts=0x7fffffffab20) at /home/jed/src/libCEED/examples/fluids/src/setupts.c:510
#26 0x000055555555b951 in main (argc=<optimized out>, argv=<optimized out>) at /home/jed/src/libCEED/examples/fluids/navierstokes.c:278
(gdb) f 8
#8  0x00007ffff58258c1 in CeedOperatorLinearAssembleQFunctionBuildOrUpdate (op=op@entry=0x55559ad019f0, assembled=assembled@entry=0x7fffffff9e80, rstr=rstr@entry=0x7fffffff9e88, request=0x7ffff5b62928 <ceed_request_immediate>) at /home/jed/src/libCEED/interface/ceed-preconditioning.c:1376
1376	      CeedCall(op->LinearAssembleQFunction(op, &assembled_vec, &assembled_rstr, request));
(gdb) p op->num_elem
$1 = 0
```

Comment (jedbrown) [reply interface/ceed-preconditioning.c:515]: Here's a reproducer at your HEAD that doesn't need a PETSc branch (like the above trace). You can replace `alacritty` with your preferred terminal (`xterm`, etc.) to get a debugger on both ranks in order to explore.
```console
$ mpiexec -n 2 alacritty -e gdb -ex r --args build/fluids-navierstokes -options_file examples/fluids/newtonianwave.yaml -{ts,snes}_monitor -petscpartitioner_type simple -dm_view -amat_type shell -ceed /gpu/cuda
```

Comment (jeremylt) [reply interface/ceed-preconditioning.c:515]: Woops, I see what the issue is

**Review by jedbrown — APPROVED** [review]
My cases work now. Thanks!

**Pull request merged**